File: src/main/java/vocadb/notification/reader/service/NotificationService.java
package vocadb.notification.reader.service;

import java.time.LocalDateTime;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.tuple.Pair;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;
import vocadb.notification.reader.client.VocaDbClient;
import vocadb.notification.reader.client.model.song.SongType;
import vocadb.notification.reader.client.model.user.Inbox;
import vocadb.notification.reader.client.model.user.UserMessageContract;
import vocadb.notification.reader.client.query.LanguagePreference;
import vocadb.notification.reader.client.query.OptionalFields;
import vocadb.notification.reader.service.dto.NotificationType;
import vocadb.notification.reader.service.dto.PV;
import vocadb.notification.reader.service.dto.SongNotificationType;
import vocadb.notification.reader.service.dto.Tag;
import vocadb.notification.reader.service.dto.notification.AlbumNotification;
import vocadb.notification.reader.service.dto.notification.ArtistNotification;
import vocadb.notification.reader.service.dto.notification.EventNotification;
import vocadb.notification.reader.service.dto.notification.Notification;
import vocadb.notification.reader.service.dto.notification.ReportNotification;
import vocadb.notification.reader.service.dto.notification.SongNotification;
import vocadb.notification.reader.service.dto.notification.UnknownNotification;
import static java.util.Collections.emptyList;
import static vocadb.notification.reader.service.dto.NotificationType.UNKNOWN;

@Component
@RequiredArgsConstructor
public class NotificationService {
    private final ObjectProvider<VocaDbClient> clientProvider;

    public Mono<Void> deleteNotifications(
            int userId,
            Collection<Integer> notificationIds,
            List<String> authCookies
    ) {
        var client = clientProvider.getObject();
        client.setCookies(authCookies);
        return Mono.fromCompletionStage(client.userApi().deleteMessages(userId, notificationIds));
    }

    public CompletableFuture<Pair<Collection<Notification>, Integer>> loadNotifications(
            int userId,
            int startOffset,
            int maxResults,
            List<String> authCookies,
            LanguagePreference languagePreference
    ) {
        var client = clientProvider.getObject();
        client.setCookies(authCookies);

        return client.userApi()
                .getMessages(userId, Inbox.NOTIFICATIONS, null, null, startOffset, maxResults, true)
                .toCompletableFuture()
                .thenApply(r -> {
                    Integer totalCount = r.totalCount();
                    if (r.items() == null) {
                        return Pair.of(emptyList(), 0);
                    }

                    List<Notification> msgStream = Stream.ofNullable(r.items()).flatMap(Collection::stream)
                            .map(msg -> client.userApi().getMessage(msg.id()).toCompletableFuture())
                            .map(msg -> toNotification(msg, client, languagePreference))
                            .map(CompletableFuture::join)
                            .collect(Collectors.toList());

                    return Pair.of(msgStream, totalCount);
                });
    }

    private CompletableFuture<? extends Notification> toNotification(
            CompletableFuture<UserMessageContract> message,
            VocaDbClient client,
            LanguagePreference languagePreference
    ) {
        return message.thenComposeAsync(msg -> {
            Pair<NotificationType, Integer> detection = detectNotificationType(msg);
            NotificationType type = detection.getLeft();
            Integer payloadId = detection.getRight();

            String originalBody = msg.body();
            String body = originalBody == null ? "<empty body>" : originalBody;

            CompletableFuture<? extends Notification> notification;
            switch (type) {
                case SONG:
                    notification = createSongNotification(
                            client,
                            msg.id(),
                            msg.subject(),
                            body,
                            payloadId,
                            languagePreference
                    );
                    break;
                case ARTIST:
                    notification =
                            CompletableFuture.completedFuture(new ArtistNotification(msg.id(), msg.subject(), body));
                    break;
                case EVENT:
                    notification =
                            CompletableFuture.completedFuture(new EventNotification(msg.id(), msg.subject(), body));
                    break;
                case ALBUM:
                    notification =
                            CompletableFuture.completedFuture(new AlbumNotification(msg.id(), msg.subject(), body));
                    break;
                case REPORT:
                    notification =
                            CompletableFuture.completedFuture(new ReportNotification(msg.id(), msg.subject(), body));
                    break;
                default:
                    notification =
                            CompletableFuture.completedFuture(new UnknownNotification(msg.id(), msg.subject(), body));
                    break;
            }

            return notification;
        });
    }

    private CompletableFuture<SongNotification> createSongNotification(
            VocaDbClient client,
            int id,
            String subject,
            String body,
            int songId,
            LanguagePreference languagePreference
    ) {
        return client.songApi()
                .getSongById(songId, List.of(OptionalFields.TAGS, OptionalFields.PVS), languagePreference)
                .toCompletableFuture()
                .thenApplyAsync(song -> {
                    SongNotificationType type = subject.contains("tagged")
                            ? SongNotificationType.TAGGED
                            : SongNotificationType.NEW;

                    String artist = song.artistString();
                    String title = song.name();
                    LocalDateTime publishDate = song.publishDate();
                    SongType songType = song.songType();
                    List<Tag> tags = Stream.ofNullable(song.tags()).flatMap(Collection::stream)
                            .map(Tag::of)
                            .sorted(Comparator.comparing(Tag::count)
                                    .thenComparing(Tag::name))
                            .collect(Collectors.toList());
                    List<PV> pvs = Stream.ofNullable(song.pvs()).flatMap(Collection::stream)
                            .map(PV::of)
                            .sorted(Comparator.<PV, String>comparing(pv -> pv.service().name())
                                    .thenComparing(pv -> pv.pvType().name()))
                            .collect(Collectors.toList());

                    return new SongNotification(
                            id,
                            subject,
                            body,
                            type,
                            songId,
                            songType,
                            tags,
                            pvs,
                            title,
                            artist,
                            publishDate
                    );
                });
    }

    @SuppressWarnings("ReturnCount")
    private Pair<NotificationType, Integer> detectNotificationType(UserMessageContract notification) {
        Pair<NotificationType, Integer> unknown = Pair.of(UNKNOWN, 0);
        String body = notification.body();
        if (body == null) {
            return unknown; // must never happen, because body is fetched
        }

        Pattern pattern = Pattern.compile(".*\\(https?://vocadb.net/(\\w+)/(\\d+)\\).*");
        Matcher matcher = pattern.matcher(body);
        if (!matcher.find()) {
            return unknown;
        }

        String typeStr = matcher.group(1);
        NotificationType type = NotificationType.of(typeStr);
        if (type == UNKNOWN) {
            return unknown;
        }

        String id = matcher.group(2);
        int matchedId = Integer.parseInt(id);
        return Pair.of(type, matchedId);
    }
}

File: src/main/java/vocadb/notification/reader/service/NotificationService.kt
package vocadb.notification.reader.service

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.future.asDeferred
import kotlinx.coroutines.future.await
import org.springframework.beans.factory.ObjectProvider
import org.springframework.stereotype.Service
import vocadb.notification.reader.client.VocaDbClient
import vocadb.notification.reader.client.api.enums.LanguagePreference
import vocadb.notification.reader.client.api.enums.OptionalFields
import vocadb.notification.reader.client.model.Inbox
import vocadb.notification.reader.client.model.PVContract
import vocadb.notification.reader.client.model.PvService
import vocadb.notification.reader.client.model.PvType
import vocadb.notification.reader.client.model.SongType
import vocadb.notification.reader.client.model.TagUsageForApiContract
import vocadb.notification.reader.client.model.UserMessageContract
import java.time.LocalDateTime

private typealias PayloadId = Long
private typealias TotalCount = Long

@Service
class NotificationService(val clientProvider: ObjectProvider<VocaDbClient>) {

    suspend fun deleteNotifications(userId: Long, notificationIds: Collection<Long>, authCookies: List<String>) {
        val client = clientProvider.getObject()
        client.cookies = authCookies

        client.userApi.deleteMessages(userId, notificationIds)
    }

    suspend fun loadNotifications(
        userId: Long,
        startOffset: Long,
        maxResults: Long,
        authCookies: List<String>,
        languagePreference: LanguagePreference
    ): Pair<Collection<Notification>, TotalCount> {
        val rawNotifications = loadRawNotifications(userId, startOffset, maxResults, authCookies)
        val notifications = loadNotificationsData(rawNotifications.notifications, authCookies, languagePreference)

        return Pair(notifications, rawNotifications.totalCount)
    }

    private suspend fun loadRawNotifications(
        userId: Long,
        startOffset: Long,
        maxResults: Long,
        authCookies: List<String>
    ): RawNotificationsResult {
        val client = clientProvider.getObject()
        client.cookies = authCookies

        val messages = client.userApi.getMessages(
            inbox = Inbox.NOTIFICATIONS,
            getTotalCount = true,
            id = userId,
            start = startOffset,
            maxResults = maxResults
        ).await()

        if (messages.items.isNullOrEmpty()) return RawNotificationsResult(emptyList(), 0)

        val notifications = messages.items.map { client.userApi.getMessage(it.id).asDeferred() }.awaitAll()
        return RawNotificationsResult(notifications, messages.totalCount!!)
    }

    private suspend fun loadNotificationsData(
        notifications: Collection<UserMessageContract>,
        authCookies: List<String>,
        languagePreference: LanguagePreference
    ): Collection<Notification> {
        if (notifications.isEmpty()) return emptyList()

        val client = clientProvider.getObject()
        client.cookies = authCookies

        return notifications
            .map { n ->
                val (type, matchedId) = detectNotificationType(n)
                val originalBody = n.body ?: "<empty body>"

                when (type) {
                    NotificationType.SONG -> createSongNotification(
                        client,
                        n.id,
                        n.subject,
                        originalBody,
                        matchedId,
                        languagePreference
                    )
                    NotificationType.ARTIST -> ArtistNotification(n.id, n.subject, originalBody)
                    NotificationType.EVENT -> EventNotification(n.id, n.subject, originalBody)
                    NotificationType.ALBUM -> AlbumNotification(n.id, n.subject, originalBody)
                    NotificationType.REPORT -> ReportNotification(n.id, n.subject, originalBody)
                    NotificationType.UNKNOWN -> UnknownNotification(n.id, n.subject, originalBody)
                }
            }
    }

    private suspend fun createSongNotification(
        client: VocaDbClient,
        id: Long,
        subject: String,
        body: String,
        songId: Long,
        languagePreference: LanguagePreference
    ): SongNotification {
        val song = client.songApi
            .getSongById(songId, listOf(OptionalFields.TAGS, OptionalFields.PVS), languagePreference)
            .await()

        val type = when {
            subject.contains("tagged") -> SongNotificationType.TAGGED
            else -> SongNotificationType.NEW
        }

        val artist = song.artistString
        val title = song?.name
        val publishDate = song?.publishDate
        val songType = song.songType
        val tags = (song.tags ?: emptyList())
            .map(Tag.Companion::of)
            .sortedWith(compareBy<Tag> { it.count }
                .thenBy { it.name }
            )
        val pvs = (song.pvs ?: emptyList())
            .map(PV.Companion::of)
            .sortedWith(compareBy<PV> { it.service.value }
                .thenBy { it.pvType }
            )

        return SongNotification(id, subject, body, type, songId, songType, title, artist, tags, pvs, publishDate)
    }

    private fun detectNotificationType(notification: UserMessageContract): Pair<NotificationType, PayloadId> {
        val unknown = Pair(NotificationType.UNKNOWN, 0L)

        val pattern = Regex(""".*\(https?://vocadb.net/(\w+)/(\d+)\).*""")
        val match = pattern.find(notification.body!!) ?: return unknown

        val type = NotificationType.fromCode(match.groupValues[1])
        if (type == NotificationType.UNKNOWN) return unknown

        val matchedId = match.groupValues[2].toLong()
        return Pair(type, matchedId)
    }
}

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "notificationType")
@JsonSubTypes(
    JsonSubTypes.Type(value = AlbumNotification::class, name = "AlbumNotification"),
    JsonSubTypes.Type(value = ArtistNotification::class, name = "ArtistNotification"),
    JsonSubTypes.Type(value = EventNotification::class, name = "EventNotification"),
    JsonSubTypes.Type(value = ReportNotification::class, name = "ReportNotification"),
    JsonSubTypes.Type(value = SongNotification::class, name = "SongNotification"),
    JsonSubTypes.Type(value = UnknownNotification::class, name = "UnknownNotification")
)
sealed class Notification(
    val id: Long,
    val originalSubject: String,
    val originalBody: String
)

class AlbumNotification(
    id: Long,
    originalSubject: String,
    originalBody: String
) : Notification(id, originalSubject, originalBody)

class ArtistNotification(
    id: Long,
    originalSubject: String,
    originalBody: String
) : Notification(id, originalSubject, originalBody)

class EventNotification(
    id: Long,
    originalSubject: String,
    originalBody: String
) : Notification(id, originalSubject, originalBody)

class ReportNotification(
    id: Long,
    originalSubject: String,
    originalBody: String
) : Notification(id, originalSubject, originalBody)

class SongNotification(
    id: Long,
    originalSubject: String,
    originalBody: String,
    val type: SongNotificationType,
    val songId: Long,
    val songType: SongType,
    val title: String? = null,
    val artist: String? = null,
    val tags: Collection<Tag>,
    val pvs: Collection<PV>,
    val releaseDate: LocalDateTime? = null
) : Notification(id, originalSubject, originalBody)

enum class SongNotificationType {
    TAGGED,
    NEW
}

data class PV(
    val id: Long,
    val pvType: PvType,
    val service: PvService,
    val url: String,
    val name: String,
    val disabled: Boolean,
    val author: String? = null,
    val publishDate: LocalDateTime? = null,
    val pvId: String? = null,
    val thumbUrl: String? = null,
    val timestamp: String? = null
) {
    companion object {
        fun of(pv: PVContract): PV {
            val timestamp = if (pv.service == PvService.PIAPRO && pv.extendedMetadata?.json != null) {
                pv.extendedMetadata.json["Timestamp"]
            } else {
                null
            }

            return PV(
                pv.id,
                pv.pvType,
                pv.service,
                pv.url,
                pv.name,
                pv.disabled,
                pv.author,
                pv.publishDate,
                pv.pvId,
                pv.thumbUrl,
                timestamp
            )
        }
    }
}

data class Tag(
    val count: Long,
    val id: Long,
    val name: String,
    val categoryName: String? = null
) {
    companion object {
        fun of(tag: TagUsageForApiContract): Tag {
            return Tag(tag.count, tag.tag.id, tag.tag.name, tag.tag.categoryName)
        }
    }
}

class UnknownNotification(
    id: Long,
    originalSubject: String,
    originalBody: String
) : Notification(id, originalSubject, originalBody)

data class RawNotificationsResult(
    val notifications: Collection<UserMessageContract>,
    val totalCount: Long
)

enum class NotificationType(val code: String) {
    SONG("S"),
    ARTIST("Ar"),
    EVENT("E"),
    ALBUM("Al"),
    REPORT("R"),
    UNKNOWN("~");

    companion object {
        @JvmStatic
        fun fromCode(code: String): NotificationType = when (code) {
            SONG.code -> SONG
            ARTIST.code -> ARTIST
            EVENT.code -> EVENT
            ALBUM.code -> ALBUM
            REPORT.code -> REPORT
            else -> UNKNOWN
        }

    }
}
